txnsync/mainloop.go

```diff
@@ -0,0 +1,436 @@
+// for non-relays.
+func (s *syncState) rollOffsets() {
+	s.nextOffsetRollingCh = s.clock.TimeoutAt(s.clock.Since() + 2*s.lastBeta)
+	s.requestsOffset++
+
+	if !s.fetchTransactions {
+		return
+	}
+
+	// iterate on the outgoing peers and see if we want to send them an update as needed.
+	// note that because this function is only called for non-relays, then all the connections
+	// are outgoing.
+	peers := s.getPeers()
+	s.updatePeersRequestParams(peers)
+
+	// check when each of these peers is expected to send a message. we might want to promote a message to be sent earlier.
+	currentTimeOffset := s.clock.Since()
+	deadlineMonitor := s.clock.DeadlineMonitorAt(currentTimeOffset + sendMessagesTime)
+
+	for _, peer := range peers {
+		nextSchedule := s.scheduler.peerDuration(peer)
+		if nextSchedule == 0 {
+			// a new peer - ignore for now. This peer would get scheduled on the next new round.
+			continue
+		}
+		if currentTimeOffset+sendMessagesTime > nextSchedule {
+			// there was a message scheduled already in less than 20ms, so keep that one.
+			s.scheduler.schedulerPeer(peer, nextSchedule)
+			continue
+		}
+
+		// otherwise, send a message to that peer. Note that we're passing the `nextSchedule-s.lastBeta` as the currentTime,
+		// so that the time offset would be based on that one. ( i.e. effectively, it would retain the existing timing, and prevent
+		// the peers from getting aligned )
+		s.sendMessageLoop(nextSchedule-s.lastBeta, deadlineMonitor, []*Peer{peer})
+	}
+}
+
+func (s *syncState) getPeers() (result []*Peer) {
+	peersInfo := s.node.GetPeers()
+	updatedNetworkPeers := []interface{}{}
+	updatedNetworkPeersSync := []*Peer{}
+
+	var averageDataExchangeRate uint64
+
+	// some of the network peers might not have a sync peer, so we need to create one for these.
+	for _, peerInfo := range peersInfo {
+		if peerInfo.TxnSyncPeer == nil {
+			syncPeer := makePeer(peerInfo.NetworkPeer, peerInfo.IsOutgoing, s.isRelay, &s.config)
+			peerInfo.TxnSyncPeer = syncPeer
+			updatedNetworkPeers = append(updatedNetworkPeers, peerInfo.NetworkPeer)
+			updatedNetworkPeersSync = append(updatedNetworkPeersSync, syncPeer)
+		}
+		result = append(result, peerInfo.TxnSyncPeer)
+		averageDataExchangeRate += peerInfo.TxnSyncPeer.dataExchangeRate
+	}
+	if len(peersInfo) > 0 {
+		averageDataExchangeRate /= uint64(len(peersInfo))
+	}
+
+	// if we have any update for the transaction sync connector, the send them via
+	// a UpdatePeers call.
+	if len(updatedNetworkPeers) > 0 || len(peersInfo) > 0 {
+		s.node.UpdatePeers(updatedNetworkPeersSync, updatedNetworkPeers, averageDataExchangeRate)
+	}
+	return result
+}
+
+func (s *syncState) updatePeersRequestParams(peers []*Peer) {
+	if s.transactionPoolFull {
+		for _, peer := range peers {
+			peer.setLocalRequestParams(0, 0)
+		}
+		return
+	}
+	if s.isRelay {
+		for _, peer := range peers {
+			peer.setLocalRequestParams(0, 1)
+		}
+	} else {
+		if s.fetchTransactions {
```

Comment:
yes, you're correct. The `updatePeersRequestParams` should be called last, so it would take into account the latest value of `s.fetchTransactions`. Note that it would be very hard to trap this issue, since the configuration of s.fetchTransactions doesn't change frequently.